src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs

```diff
@@ -526,5 +558,71 @@ private bool IsV2LocalRepository(string root)
 
             return true;
         }
+
+        // Get a temp API key from nuget.org for pushing to https://nuget.smbsrc.net/
+        private async Task<string> GetSecureApiKey(
+            PackageIdentity packageIdentity,
+            string apiKey,
+            ILogger logger,
+            CancellationToken token)
+        {
+            if (string.IsNullOrEmpty(apiKey))
+            {
+                return apiKey;
+            }
+            var serviceEndpointUrl = GetServiceEndpointUrl(NuGetConstants.DefaultGalleryServerUrl,
+                string.Format(TempApiKeyServiceEndpoint, packageIdentity.Id, packageIdentity.Version));
+
+            try
+            {
+                return await _httpSource.ProcessStreamAsync(
+                    new HttpSourceRequest(
+                        () =>
+                        {
+                            var request = HttpRequestMessageFactory.Create(
+                                HttpMethod.Put,
+                                serviceEndpointUrl,
+                                new HttpRequestMessageConfiguration(
+                                    logger: logger,
+                                    promptOn403: false));
+                            request.Headers.Add(ApiKeyHeader, apiKey);
+                            request.Headers.Add(NuGetClientVersionHeader, UserAgent.UserAgentVersion);
```

Comment:
I think we should add the client version header to every HTTP call -- just like user agent.